min-base-fee: bring in op-geth code

geoknee opened the pull request:
Raising this PR to trigger monorepo CI checks and validate https://github.com/ethereum-optimism/op-geth/pull/666. This PR aims to be the minimal changeset to get existing tests passing. Note that interop tests implicitly activate jovian so we need to attach a bunch of code changes to this PR to keep those tests working.

The PR which activates the feature in full is under review here https://github.com/ethereum-optimism/optimism/pull/16852

Co-authored-by: William Law <williamlaw.wtl@gmail.com>
Co-authored-by: Niran Babalola <niran@niran.org>

geoknee committed:
boilerplate for Jovian fork

geoknee committed:
update op-geth version

brings in min-base-fee code

geoknee committed:
changes to get current tests passing

geoknee committed:
kurtosis devnet: use temporary image for op-geth

Matches the version currently in go.mod. Can be reverted to default once the op-geth release is made.

We need to do this because interop implies jovian, so we need a jovian compatible op-geth to run with the longer extradata from genesis.

geoknee commented on kurtosis-devnet/interop.yaml:
```diff
@@ -47,7 +47,7 @@ optimism_package:
           el:
             type: op-geth
-            image: ""
+            image: "us-docker.pkg.dev/oplabs-tools-artifacts/images/op-geth:710a9697279600e7812c8847283eeb750d60ed8e@sha256:0e0dd748d5596ebd728b8d4691f26c4a4a6f387a834f2aa38534e192514bec2a"
```
This is necessary because interop implies jovian, so we need a jovian-compatible op-geth to run this network with. With this image running, I manually confirmed the 17 byte extra data shows up on both network ELs.

sebastianst commented on go.mod:
```diff
@@ -302,7 +302,7 @@ require (
 )
 
-replace github.com/ethereum/go-ethereum => github.com/ethereum-optimism/op-geth v1.101602.0-rc.1
+replace github.com/ethereum/go-ethereum => github.com/ethereum-optimism/op-geth v1.101602.0-rc.1.0.20250902174942-710a96972796
```
needs to be updated when op-geth PR is merged.

geoknee replied:
[182fabd](https://github.com/ethereum-optimism/optimism/pull/17307/commits/182fabd5a2d9ddbdcc529ce46ff30eed55916420)

sebastianst commented on op-node/rollup/attributes/engine_consolidate.go:
```diff
@@ -125,7 +135,17 @@ func checkEIP1559ParamsMatch(opCfg *params.OptimismConfig, attrParams *eth.Bytes
+			}
+		} else {
+			bd, be = eip1559.DecodeHoloceneExtraData(blockExtraData)
+		}
```
See [this op-geth PR comment](https://github.com/ethereum-optimism/op-geth/pull/666/files#r2318796413), then we can use `DecodeOptimismExtraData` here.

The `len(blockExtraData) > 0` check below may be come unnecessary then, because we'd then perform the appropriate length checks depending on the active fork.

geoknee replied:
[899bb41](https://github.com/ethereum-optimism/optimism/pull/17307/commits/899bb4169cb86a206491561b73ca0abf659b3d2f)

sebastianst commented on op-node/rollup/attributes/engine_consolidate.go:
```diff
@@ -97,7 +97,8 @@ func checkParentBeaconBlockRootMatch(attrRoot, blockRoot *common.Hash) error {
 }
 
-func checkEIP1559ParamsMatch(opCfg *params.OptimismConfig, attrParams *eth.Bytes8, blockExtraData []byte) error {
+func checkEIP1559ParamsMatch(opCfg *params.OptimismConfig, attrParams *eth.Bytes8, blockExtraData []byte, minBaseFee uint64, isJovian bool) error {
```
* We should rename this function to something like `checkExtraDataParamsMatch` because it's now also checking the min base fee.
* We could pass it the rollup config and block timestamp, so that we can then pass down the rollup config to `ValidateOptimismExtraData` (see also [this review comment](https://github.com/ethereum-optimism/op-geth/pull/666/files#r2318816730))
* I suggest that we make the `minBaseFee` a `*uint64` ptr so we encode explicitly whether it's set or not (also mentioned in other review comments)
* We should rearrange the parameters here so that we first list all attribute params together, then the `blockExtraData`

Result would be a function signature like
```go
func checkEIP1559ParamsMatch(rollupCfg *rollup.Config, ts uint64, attrParams *eth.Bytes8, attrMinBaseFee *uint64, blockExtraData []byte) error
```

geoknee replied:
[e19aa9e](https://github.com/ethereum-optimism/optimism/pull/17307/commits/e19aa9e5befde56d962358b505821fad58d84298)

sebastianst commented on op-node/rollup/attributes/engine_consolidate.go:
```diff
@@ -109,10 +110,19 @@ func checkEIP1559ParamsMatch(opCfg *params.OptimismConfig, attrParams *eth.Bytes
+				// This can happen if the unsafe chain contains invalid (in particular, empty) extraData while Holocene
+				// is active. The extraData field of blocks from sequencer gossip isn't currently checked during import.
+				return fmt.Errorf("invalid block extraData: %w", err)
+			}
```
Could use `eip1559.ValidateOptimismExtraData` here after applying other PR comments

geoknee replied:
https://github.com/ethereum-optimism/optimism/pull/17307/commits/899bb4169cb86a206491561b73ca0abf659b3d2f

sebastianst commented on op-node/rollup/attributes/engine_consolidate_test.go:
```diff
@@ -36,6 +36,55 @@ type matchArgs struct {
+	}
+}
+
 func holoceneArgs() matchArgs {
```
would be good to follow the pattern to have `holoceneArgs` now use `jovianArgs` and then set the appropriate fields to `nil` (min base fee fields). That's also was `ecotoneArgs` etc do below.

geoknee replied:
411988d87c1ffbae74e6d254e4a420893cea29b1

sebastianst commented on op-node/rollup/attributes/engine_consolidate_test.go:
```diff
@@ -190,6 +239,7 @@ func TestAttributesMatch(t *testing.T) {
 
 	rollupCfgPreCanyon := &rollup.Config{CanyonTime: &futureTime, ChainOpConfig: defaultOpConfig}
 	rollupCfgPreIsthmus := &rollup.Config{CanyonTime: &pastTime, IsthmusTime: &futureTime, ChainOpConfig: defaultOpConfig}
+	rollupCfgPostJovian := &rollup.Config{CanyonTime: &pastTime, JovianTime: &futureTime, ChainOpConfig: defaultOpConfig}
```
We probably also want to set the isthmus time to `pastTime` if we do the other changes that I propose for consolidation.

sebastianst commented on op-node/rollup/attributes/engine_consolidate_test.go:
Feels like we're missing a few new test cases for invalid Jovian consolidation, e.g. eip1559 params and everything else valid, but then missing min base fee either in the attribs or block, or both present but of different values.

geoknee replied:
bcf36fdae7a199402b44f7bb0e884ece1c6da471

sebastianst commented on op-node/rollup/derive/payload_util.go:
```diff
@@ -90,7 +90,14 @@ func PayloadToSystemConfig(rollupCfg *rollup.Config, payload *eth.ExecutionPaylo
+	} else if rollupCfg.IsHolocene(uint64(payload.Timestamp)) {
 		if err := eip1559.ValidateHoloceneExtraData(payload.ExtraData); err != nil {
 			return eth.SystemConfig{}, err
 		}
```
We may be able to improve this to use the op-geth `eip1559.ValidateOptimismExtraData` and `eip1559.DecodeOptimismExtraData` functions (after the other changes I propose). This becomes more maintainable over time rather than duplicating this logic here, in consolidation, and other places.

geoknee replied:
[425d3cc](https://github.com/ethereum-optimism/optimism/pull/17307/commits/425d3cc4508c41ee86fc1749c0ff85039cac0339)

sebastianst commented on op-node/rollup/interop/indexing/attributes.go:
```diff
@@ -54,7 +54,7 @@ func AttributesToReplaceInvalidBlock(invalidatedBlock *eth.ExecutionPayloadEnvel
 	// So we translate it here.
 	extraData := invalidatedBlock.ExecutionPayload.ExtraData
-	denominator, elasticity := eip1559.DecodeHoloceneExtraData(extraData)
+	denominator, elasticity, minBaseFee := eip1559.DecodeJovianExtraData(extraData)
```
Does this rely on the best-effort decoding logic?

geoknee replied:
Hmm yes I think it does. And now we changed `PayloadAttributes.MinBaseFee` to a pointer, this now looks wrong since we will pass in a non nil pointer pre Jovian :-/

geoknee replied:
I will fix this by getting the op-geth library code to return a `*uint64` for `minBaseFee`

sebastianst commented on op-program/client/l2/engineapi/block_processor.go:
```diff
@@ -61,7 +61,11 @@ func NewBlockProcessorFromPayloadAttributes(provider BlockDataProvider, parent c
 			e = provider.Config().ElasticityMultiplier()
 		}
-		header.Extra = eip1559.EncodeHoloceneExtraData(d, e)
+		if provider.Config().IsJovian(header.Time) {
```
You can use a `atts.MinBaseFee != nil` check here if we change the field to a `*uint64` ptr.

I'm wondering if we should also introduce a function `eth.EncodeOptimismExtraData(cfg *types.ChainConfig, attrs *eth.PayloadAttributes)` , so we define this (and surrounding) logic in an isolated function that can be better maintained.

geoknee replied:
I added that check and we return an error if Jovian is active but the min base fee is nil. 

I had also wondered about a generalised encoding function. I will have a play with that.

geoknee replied:
https://github.com/ethereum-optimism/optimism/pull/17307/commits/899bb4169cb86a206491561b73ca0abf659b3d2f#diff-5a65f023808ce10cb2c99c87c423d10cf72d21deeeec2cab25676c7602ad11ec

sebastianst commented on op-program/client/l2/engineapi/l2_engine_api.go:
```diff
@@ -355,19 +356,20 @@ func (ea *L2EngineAPI) NewPayloadV3(ctx context.Context, params *eth.ExecutionPa
-		if err := eip1559.ValidateHoloceneExtraData(params.ExtraData); err != nil {
-			return &eth.PayloadStatusV1{Status: eth.ExecutionInvalid}, engine.UnsupportedFork.With(errors.New("invalid holocene extraData post-holocene"))
+	if cfg.IsOptimism() {
+		if err := eip1559.ValidateOptimismExtraData(cfg, uint64(params.Timestamp), params.ExtraData); err != nil {
```
here we see how nice it is to have the logic isolated in the separate `eip1559.ValidateOptimismExtraData` function.

sebastianst commented on op-program/client/l2/engineapi/l2_engine_api_test.go:
```diff
@@ -160,25 +163,53 @@ func newStubBackend(t *testing.T) *stubCachingBackend {
+	return createGenesisWithForkTime(0)
+}
+
+func createGenesisWithForkTime(forkTime uint64) *core.Genesis {
```
shouldn't this be named `...WithForkTimeOffset`?

geoknee replied:
Yes, b81ce85060dbbc30dea846545f738a073556c066

sebastianst commented on op-program/client/l2/engineapi/l2_engine_api_test.go:
```diff
@@ -50,6 +48,7 @@ func TestNewPayloadV4(t *testing.T) {
 		eip1559Params := eth.Bytes8([]byte{0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8})
+		minBaseFee := uint64(1e9)
 		gasLimit := eth.Uint64Quantity(4712388)
 		result, err := engineAPI.ForkchoiceUpdatedV3(context.Background(), &eth.ForkchoiceState{
```
I believe that the engine API check for op-program lag behind the latest update that we did for op-geth. We should track for follow-up work to harden the replicated engine api checks (maybe we can even reuse the extracted check functions from op-geth so we don't need to duplicate).

One result of this may be that we need to set the `minBaseFee` field to `nil` if `Jovian` is not active.

sebastianst replied:
Tracking with https://github.com/ethereum-optimism/optimism/issues/17336

sebastianst commented on op-program/client/tasks/deposits_block.go:
```diff
@@ -145,7 +145,12 @@ func blockToDepositsOnlyAttributes(cfg *rollup.Config, block *types.Block, outpu
 		GasLimit:              &gasLimit,
 	}
-	if cfg.IsHolocene(block.Time()) {
+	if cfg.IsJovian(block.Time()) {
+		d, e, m := eip1559.DecodeJovianExtraData(block.Extra())
+		eip1559Params := eth.Bytes8(eip1559.EncodeHolocene1559Params(d, e))
+		attrs.EIP1559Params = &eip1559Params
+		attrs.MinBaseFee = m
+	} else if cfg.IsHolocene(block.Time()) {
 		d, e := eip1559.DecodeHoloceneExtraData(block.Extra())
 		eip1559Params := eth.Bytes8(eip1559.EncodeHolocene1559Params(d, e))
 		attrs.EIP1559Params = &eip1559Params
```
may be able to use `DecodeOptimismExtraData` here

geoknee replied:
[899bb41](https://github.com/ethereum-optimism/optimism/pull/17307/commits/899bb4169cb86a206491561b73ca0abf659b3d2f)

sebastianst replied:
It's still not used here?

geoknee replied:
Sorry, missed that the fist time. Got it here [ef333fd](https://github.com/ethereum-optimism/optimism/pull/17307/commits/ef333fd6bacf7e325823e22abad35f3dfaf06e5d)

sebastianst commented on op-service/eth/types.go:
```diff
@@ -519,6 +519,8 @@ type PayloadAttributes struct {
 	// EIP-1559 parameters, to be specified only post-Holocene
 	EIP1559Params *Bytes8 `json:"eip1559Params,omitempty"`
+	// MinBaseFee is the minimum base fee, to be specified only post-Jovian
+	MinBaseFee uint64 `json:"minBaseFee,omitempty"`
```
see other comments, I suggest we make this a pointer so we can explicitly encode that it's not set.

sebastianst commented on op-service/flags/flags.go:
```diff
@@ -22,6 +22,7 @@ const (
 	PectraBlobScheduleOverrideFlagName = "override.pectrablobschedule"
 	IsthmusOverrideFlagName            = "override.isthmus"
 	InteropOverrideFlagName            = "override.interop"
+	JovianOverrideFlagName             = "override.jovian"
```
Also needs to be added below to `CLIFlags` return slice

geoknee replied:
e3e086925abddd98d8d08ada2d77f72bb2cc465b

geoknee committed:
bump op-geth

geoknee committed:
update expected error message in TestCheckEIP1559ParamsMatch

geoknee committed:
update to take account of PayloadAttributes.MinBaseFee being a *uint64

geoknee committed:
op-program fixes

geoknee committed:
fix computePayloadId

geoknee committed:
use static error

geoknee committed:
update op-geth image for interop kt-devnet

geoknee committed:
createGenesisWithForkTime => createGenesisWithForkTimeOffset

geoknee committed:
fully add interop override flag to op-service

geoknee committed:
refactor: define holoceneArgs in terms of jovianArgs

geoknee committed:
add test cases

geoknee committed:
MinBaseFeeExtraData -> JovianExtraData

geoknee commented on op-program/client/l2/test/miner.go:
```diff
@@ -121,6 +121,7 @@ func (m *Miner) Fork(t *testing.T, blockNumber uint64, attrs *eth.PayloadAttribu
 			NoTxPool:              true,
 			GasLimit:              &gasLimit,
 			EIP1559Params:         &eip1559Params,
+			MinBaseFee:            nil,
```
Flagging this -- is it safe to just use `nil` here?

sebastianst replied:
better set it conditionally on `m.backend.Config().IsJovian(uint64(attrs.Timestamp))`

geoknee replied:
[a64bffd](https://github.com/ethereum-optimism/optimism/pull/17307/commits/a64bffdc115916bfb32dde63ed5c4e45043e6648)

geoknee replied:
Hmm actually what we are doing here seems to be using default values when `attrs=nil`. So there is nothing to set it to in that case.

I'll check with the proofs team about what default makes the most sense here.

geoknee committed:
bump op-geth version, make use of eip1559.DecodeOptimismExtraData

geoknee committed:
checkEIP1559ParamsMatch -> checkExtraDataParamsMatch

geoknee committed:
bump op-geth

geoknee committed:
go mod tidy

geoknee committed:
use eip1559.Validate/DecodeOptimismExtraData in PayloadToSystemConfig

geoknee committed:
bring interop devnet op-geth in line with go.mod op-geth

geoknee committed:
fix bug (pointer comparison)

geoknee committed:
fix test

geoknee committed:
go mod tidy

geoknee committed:
Merge remote-tracking branch 'origin/develop' into gk/op-geth-cmb2

sebastianst commented on op-node/rollup/attributes/engine_consolidate.go:
```diff
@@ -109,30 +109,35 @@ func checkEIP1559ParamsMatch(opCfg *params.OptimismConfig, attrParams *eth.Bytes
 
-		bd, be := eip1559.DecodeHoloceneExtraData(blockExtraData)
+		// Decode block parameters and check for mismatch
+		err := eip1559.ValidateOptimismExtraData(cfg, blockTimestamp, blockExtraData)
+		if err != nil {
+			return fmt.Errorf("invalid block extraData: %w", err)
+		}
```
I suggest we move this check to the top of `checkExtraDataParamsMatch` so it always runs, also for pre-Holocene blocks.

geoknee replied:
[c300b50](https://github.com/ethereum-optimism/optimism/pull/17307/commits/c300b50b8a288b1b4bc3fbcb20e83762eaf14b73)

geoknee replied:
I ended up reverting this [b54581d](https://github.com/ethereum-optimism/optimism/pull/17307/commits/b54581dfa78ce5d9b347730030e054425fe3eb43), I didn't want to change the existing behaviour (it broke existing tests). Happy to look again in a followup PR>

sebastianst replied:
Interesting, worth following up on.

geoknee replied:
https://github.com/ethereum-optimism/optimism/issues/17337

geoknee committed:
tag issue in TODO

geoknee committed:
consolidate: validate extra data even pre Holocene and fix test

geoknee committed:
fix test miner to set minbasefee correctly

geoknee committed:
use eip1559.DecodeOptimismExtraData in blockToDepositsOnlyAttributes

geoknee committed:
bump op-geth

geoknee committed:
go mod tidy

geoknee committed:
bump op-geth in kt devnet

geoknee committed:
only validate extra data from holocene

sebastianst reviewed (approved):
Approving, contingent on updating the op-geth reference to a tagged version once the op-geth PR is merged.

geoknee committed:
bump op-geth to tagged RC

geoknee committed:
use pre-Jovian value in test miner.Fork

geoknee committed:
op-program test miner: set min base fee to a stub (valid) value if Jovian is active

Inphi commented on op-program/client/l2/engineapi/l2_engine_api.go:
```diff
@@ -355,19 +358,20 @@ func (ea *L2EngineAPI) NewPayloadV3(ctx context.Context, params *eth.ExecutionPa
-	if ea.config().IsHolocene(uint64(params.Timestamp)) {
-		if err := eip1559.ValidateHoloceneExtraData(params.ExtraData); err != nil {
-			return &eth.PayloadStatusV1{Status: eth.ExecutionInvalid}, engine.UnsupportedFork.With(errors.New("invalid holocene extraData post-holocene"))
+	if cfg.IsOptimism() {
```
nit: `cfg.IsOptimism()` is a no-op. We're always running op rules inside the program.

geoknee replied:
[afe7721](https://github.com/ethereum-optimism/optimism/pull/17307/commits/afe77210967ab498d375953edcfb22184b3377d7)

Inphi reviewed (approved):
Proofs changes look good to me. It's nice to have the header codec rules fully handled by op-geth.

geoknee committed:
remove redundant check on isOptimism

We are always running op rules inside the program

geoknee committed:
Co-authorship credit

This PR contains work that was cherry-picked from https://github.com/ethereum-optimism/optimism/pull/16852.
This commit will mean they are credited with co-authorship.

Co-authored-by: William Law <williamlaw.wtl@gmail.com>
Co-authored-by: Niran Babalola <niran@niran.org>
